sklearn/manifold/_t_sne.py

```diff
@@ -745,6 +747,28 @@ class TSNE(BaseEstimator):
     (4, 2)
     """
 
+    _parameter_constraints = {
+        "n_components": [Interval(Integral, 1, None, closed="left")],
+        "perplexity": [Real],
+        "early_exaggeration": [Interval(Real, 1, None, closed="left")],
+        "learning_rate": [
+            StrOptions({"auto", "warn"}),
+            Interval(Real, 0, None, closed="neither"),
+        ],
+        "n_iter": [Interval(Integral, 250, None, closed="left")],
+        "n_iter_without_progress": [Integral],
```

Comment:
It doesn't make sense to have `n_iter_without_progress < -1`. Technically `n_iter_without_progress=-1` would really only be a debug convenience at best since then the check `i - best_iter > n_iter_without_progress` would trivially be `True`.
```suggestion
        "n_iter_without_progress": [Interval(Integer, -1, None, closed="left")],
```